contracts/twap/CumulativeTwap.sol

```diff
@@ -91,33 +106,36 @@ contract CumulativeTwap is BlockContext {
         (Observation memory beforeOrAt, Observation memory atOrAfter) = _getSurroundingObservations(targetTimestamp);
         uint256 targetCumulativePrice;
 
-        // case1. not enough historical data or just enough (`==` case)
-        if (targetTimestamp <= beforeOrAt.timestamp) {
-            targetTimestamp = beforeOrAt.timestamp;
+        // case1. left boundary
+        if (targetTimestamp == beforeOrAt.timestamp) {
             targetCumulativePrice = beforeOrAt.priceCumulative;
         }
-        // case2. the latest data is older than or equal the request
-        else if (atOrAfter.timestamp <= targetTimestamp) {
-            targetTimestamp = atOrAfter.timestamp;
+        // case2. right boundary
+        else if (atOrAfter.timestamp == targetTimestamp) {
             targetCumulativePrice = atOrAfter.priceCumulative;
         }
         // case3. in the middle
         else {
-            uint256 observationTimeDelta = atOrAfter.timestamp - beforeOrAt.timestamp;
+            // it implies beforeOrAt = observations[currentObservationIndex]
+            // which means there's no atOrAfter
+            if (atOrAfter.timestamp == 0) {
+                atOrAfter.priceCumulative =
+                    beforeOrAt.priceCumulative +
+                    (beforeOrAt.price * (targetTimestamp - beforeOrAt.timestamp));
+                atOrAfter.timestamp = targetTimestamp;
+            }
```

Comment:
we can directly return `targetCumulativePrice = atOrAfter.priceCumulative` here?

Reply:
ah right 🙏 thanks

Reply:
fixed in f64de2e